tothic/tothic.go

```diff
@@ -134,11 +134,25 @@ func BeginAuthHandler(res http.ResponseWriter, req *http.Request, toth *toth.Tot
 	http.Redirect(res, req, url, http.StatusTemporaryRedirect)
 }
 
-// GetState gets the state string associated with the given request
-// This state is sent to the provider and can be retrieved during the
-// callback.
+// GetState gets the state returned by the provider during the callback.
+// This is used to prevent CSRF attacks, see
+// http://tools.ietf.org/html/rfc6749#section-10.12
 var GetState = func(req *http.Request) string {
-	return "state"
+	params := req.URL.Query()
+
+	var state string
+
+	if params.Encode() == "" && req.Method == http.MethodPost {
+		state = req.FormValue("state")
+	} else {
+		state = params.Get("state")
+	}
+
+	if state == "" {
+		state = "state"
+	}
+
+	return state
```

Comment:
I guess we can make the fucntion clearer and use early returns, am thinking in something like:

```
    params := req.URL.Query()

    // Try to get "state" from query parameters first
    state := params.Get("state")

    // If no "state" in query parameters and it's a POST request, try form data
    if state == "" && req.Method == http.MethodPost {
        state = req.FormValue("state")
    }

    // If no "state" found, return the default value
    if state == "" {
        state = "state"
    }

    return state
```

Also avoid the use of params.Encode() to know if we have something there in the query params. WDYT?